aztec/AZ80106D/ctools/doc/ctools.1
.TH CTOOLS 1 "1996\-02\-20" "version 0.4"
.\" $Id: ctools.1,v 1.1 1997/02/20 23:29:31 jochen Rel $
.\" Copyright (c) 1995-1996  Jochen Metzinger.
.\"
.\" Permission is granted to make and distribute verbatim copies of
.\" this manual provided the copyright notice and this permission notice
.\" are preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\"
.\" Permission is granted to copy and distribute translations of this
.\" manual into another language, under the above conditions for modified
.\" versions, except that this permission notice may be included in
.\" translations instead of in the original English.
.\"
.SH NAME
ctools \- manipulate c64/c128 CP/M disk images
.SH SYNOPSIS
.B ctools
.RB [ \-hi ]
.RB [ \-u
.IR user ]
.RB [ \-\-help ]
.RB [ \-\-initdir ]
.RB [ \-\-user=\fIuser ]
.I image
.I command
.RI [ parameter ]
.SH DESCRIPTION
.B ctools
is a public domain program to allow Unix and MSDOS systems to read,
write and manipulate files on an c64/c128 CP/M disk image.
.SH "CP/M FILE NAMES"
.BR ctools '
CP/M file have the format
.PP
	[\fIuser\fB:\fR]\fIname\fP[\fB.\fP[\fPtype\fP]]
.PP
where
.I user
is a letter from
.B A
to
.BR P ,
.I name 
a string of up to 8 and
.I type
a string of up to 3 character.
The character 
.B ?
is a joker for one character and
.B *
for the rest of the part
.RI ( user ,
.I name
or
.IR type ).
.SH OPTIONS
.TP
.BR \-i ", " \-\-initdir
If you have passwords, directory labels or time stamps on your image
you will not be able to work with
.BR ctools .
You have to use this option to remove this entries.
.TP
.BR "-u \fIuser" ", " "\-\-user=\fIuser"
Set the default user array to
.IR user .
If you do not use this option the default user array will be
.BR 0 .
.TP
.BR \-h ", " \-\-help
print a help message on standard output and exit successfully.
.SH COMMANDS
.TP
.B c
check the directory of the
.IR image .
If the option
.B \-i
is given the cleared directory will be saved to the
.IR image .
.TP
.BR "d " "[\fIfile\fP]"
output the directory of
.IR file .
The default
.I file 
is
.BR ?:*.* .
.TP
.BR "e " "[\fIfile\fP]"
delete each
.IR file .
.TP
.BR "g " "[\fIfile\fP]"
copy each 
.I file
from the 
.I image
to the current directory.
.TP
.BR h ", " ?
print a help message on standard output and exit successfully.
.TP
.BR "m " "[\fIfile\fP]"
output a dump for each
.IR file .
.TP
.BR "p " "[\fIfile\fP]"
copy each
.I file
to the user array
.I user 
of the 
.IR image .
.TP
.BI "r " "old new"
renames 
.I old
to 
.RI new .
.TP
.BR "t " "[\fIfile\fP]"
output each file matching 
.IR file .
.TP
.B u
undelete deleted files.
.SH "SEE ALSO"
.BR d64dump (1),
.BR biosdump (1),
.BR cformat (1),
.BR d64 (5)
.SH AUTHOR
Jochen Metzinger<1metzing@informatik.uni\-hamburg.de>


aztec/AZ80106D/ctools/src/ctools.cc
//
// $Id: ctools.cc,v 1.1 1997/02/20 23:21:23 jochen Rel $
// Copyright (C) 1995-1997  Jochen Metzinger
//
// main programm of ctools
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//
// $Log: ctools.cc,v $
// Revision 1.1  1997/02/20 23:21:23  jochen
// from ctools 3.0
//

#ifndef lint
static const char rcsid[] = "$Id";
#endif /* not lint */


#include <ctype.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include "err.h"
#include "fcb.h"
#include "tools.h"


static void
usage()
{
	puts("ctools " VERSION "  manipulate c64/c128 CP/M disk images");
	puts("Copyright (c) 1995-1997  Jochen Metzinger\n");
	printf("Usage: %s\n\n", short_usage);
	puts("-i --initdir           remove SFCB, XFCB and dir labels");
	puts("-u --user user-array   set default user array (a-p)");
	puts("image                  disk image");
	puts("command                tool:");
	puts("  c                      check directory (correct with -i)");
	puts("  d [file]               show directory");
	puts("  e file                 erase file");
	puts("  g file                 get file from image");
	puts("  h,?                    show this display this help and exit");
	puts("  m file                 output dump of file");
#ifdef MSDOS
	puts("  p file                 put file onto image");
#else
	puts("  p files                put files onto image");
#endif
	puts("  r old new              rename old to new");
	puts("  t file                 output file");
	puts("  u                      unerase files\n");
	puts("-h --help              print this help, then exit");
        exit(EXIT_SUCCESS);
}

int
main (int argc, char *argv[])
{
	char command, *image;
	bool initdir = FALSE;

#ifdef MSDOS
        source = "ctools";
#else
        source = argv[0];
#endif
	short_usage = "ctools [--help] [-i] [-u user-array]" \
	              " image command [options]";
	err_init();
	while (1) {
		static int c, index;
		static struct option long_options[] = {
			{"help",    no_argument,       NULL, 'h'},
			{"initdir", no_argument,       NULL, 'i'},
			{"user",    required_argument, NULL, 'u'},
			{0, 0, 0, 0},
		};

		c = getopt_long (argc, argv, "hiu:", long_options, &index);
		if (c == EOF)
			break;
		switch (c) {
			case 'h':
				usage();
				break;
			case 'i':
				initdir = TRUE;
				break;
			case 'u':
				c = toupper(*optarg);
				if ('A' <= c && c <= 'P')
					default_user_array = c - 'A';
				else
					errx("bad user array %c", c);
				break;
			case ':':
			case '?':
				errx();
				break;
			default:
				errx("unknown option %c", c);
		}
	}
	if (optind == argc)
                errx();
	image = argv[optind++];
	if (optind == argc)
                errx();
	command = argv[optind++][0];
	if (command == '\0')
		errx();
	if (islower(command))
		command = toupper(command);
	init(image, initdir);
	short_usage = NULL;
	switch (command) {
		case 'C':
			if (initdir)
				flush_dir();
			break;
		case 'D':
			print_dir((optind == argc) ? "*.*" : argv[optind]);
			break;
		case 'E':
			if (optind+1 != argc)
				errx("expected one parameter"); 
			erase_files(argv[optind]);
			break;
		case 'G':
			if (optind+1 != argc)
				errx("expected one parameter"); 
			get_files(argv[optind]);
			break;
		case 'H': case '?':
			usage();
			break;
		case 'M':
			if (optind+1 != argc)
				errx("expected one parameter"); 
			dump_files(argv[optind]);
			break;
		case 'P':
#ifdef MSDOS
			if (optind+1 != argc)
				errx("expected one parameter");
			put_files(argv[optind]);
#else
			if (optind == argc)
				errx("expected parameters"); 
			while (optind < argc) 
				put_file(argv[optind++]);
#endif
			break;
		case 'R':
			if (optind+2 != argc)
				errx("expected two parameters"); 
			rename_files(argv[optind], argv[optind+1]);
			break;
		case 'T':
			if (optind+1 != argc)
				errx("expected one parameter"); 
			print_files(argv[optind]);
			break;
		case 'U':
			if (optind != argc)
				errx("expected no parameter"); 
			unerase_files();
			break;
		default:
			errx("unknown command %c", command);
	}
	done();
	return EXIT_SUCCESS;
}
